### PR TITLE
Add InvalidBounceTerrain parameter to Projectile

### LINKED ISSUE #18742: Valid Bounce Terrain
Hi,

OpenRA supports bouncing logic for bullet projectiles. It was of course added due to Tiberian Sun game. However, the bouncing logic doesn't understand terrain types, so naturally it will bounce off water tiles. It should either explode or sink. Now I know that in the original TS game, discs bounce off water tiles which is hard to tell whether it was intentional or overlooked, knowing that TS was a rushed game.

Either way, I'd love to see additional parameter, let's say `ValidBounceTerrain` which would tell which terrain types are valid for bouncing off.

Currently, my [OpenSA](https://github.com/Dzierzan/OpenSA) mod requires that as in the original game, boucing projectiles explode when they touch water tiles. Dzierzan/OpenSA#137

Regards!

### PR DESCRIPTION
**Pull request by abc013**
Closes #18742, but adds `InvalidBounceTerrain` instead of the requested `BounceTerrain`. This is because it makes more sense to me that one wants a projectile to bounce off all tiles but a specific one (in this case: water). However, if requested, I can also add `BounceTerrain`.

Testcase: Adds `InvalidBounceTerrain: Water` to the disk thrower in TS. However, it looks quite strange when the projectile explodes on coast tiles, which is why I wouldn't recommend using it for bleed.

### CODE REVIEW

**Review by abcdefg30 — DISMISSED** [review]
Lgtm.

File: OpenRA.Mods.Common/Projectiles/Bullet.cs
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]:
```suggestion
		public readonly HashSet<string> InvalidBounceTerrain = new HashSet<string>();
```

seems to be the standard in the current code base. I guess because of performance.

Comment (abc013) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs]: Done.

**Comment by abc013** [comment]
While I was at it, I also removed the `shouldExplode` boolean and replaced it with `return` statments to prevent a tiny bit of unnecessary calculations (see second commit).

**Comment by Mailaender** [comment]
You can remove https://github.com/OpenRA/OpenRA/pull/18779/commits/8d35a84e66b611866ca130f0e9f6e7b57431b401 now.

File: OpenRA.Mods.Common/Projectiles/Bullet.cs
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs:253]: The early return above change behavior here as these values won't be set?

**Comment by phrohdoh** [comment]
Why would they remove the testcase before acquiring all the required approvals?

**Comment by abcdefg30** [comment]
Fwiw, the commit would still be accessible through the link in the comment.

**Comment by abc013** [comment]
I removed the testcase.

**Comment by phrohdoh** [comment]
> Fwiw, the commit would still be accessible through the link in the comment.

Sure, but not nearly as easy to test (cannot just checkout the PR as-is now) and review.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9130ae92e9ca14c4b377a08b78deed8332289b91...97bd7f67259064bd3c0c08f151494770971a9973)

**Pull request merged**
